File: gke_customization/gke_hrms/report/employee_punch_error/employee_punch_error.js
// Copyright (c) 2024, Gurukrupa Export and contributors
// For license information, please see license.txt

frappe.query_reports["Employee Punch Error"] = {
	"filters": [
		{
			"label": __("Company"),
			"fieldname": "company",
			"fieldtype": "Link",
			"options": 'Company',
			"reqd": 1
		},
		{
			"label": __("From Date"),
			"fieldname": "from_date",
			"fieldtype": "Date",
			"reqd": 1
		},
		{
			"label": __("To Date"),
			"fieldname": "to_date",
			"fieldtype": "Date",
			"reqd": 1
		},
		{
			"label": __("Department"),
			"fieldname": "department",
			"fieldtype": "Link",
			"options": 'Department',
			// "reqd": 1,
			"get_query": function() {
				var company = frappe.query_report.get_filter_value('company')
				return {
					"doctype": "Department",
					"filters": {"company": company}
				}
			},
		},
		{
			"label": __("Employee"),
			"fieldname": "employee",
			"fieldtype": "Link",
			"options": 'Employee'
		},	
		

	],
	onload: (report) => {
		
		report.page.add_button("Clear Filters", function() {
			window.open("/app/query-report/Employee%20Punch%20Error", "_self")
		}).addClass("btn-info");

		report.page.add_button("Generate", function() {
			var date = frappe.query_report.get_filter_value('attendance_date');
			var employeeId = frappe.query_report.get_filter_value('employee');
			frappe.query_report.set_filter_value({
				"date": date,
				"employee": employeeId
			});
		}).addClass("btn-primary");

		report.page.add_inner_button(__("Mark as Attendance"), function () {
			// Select the checked checkbox inside the datatable
			let checkedBox = document.querySelector('.dt-cell input[type="checkbox"]:checked');
		
			if (!checkedBox) {
				frappe.msgprint(__("Please select at least one row."));
				return false;
			}
		
			// Get the data-row-index from the closest dt-cell div
			let selectedCell = checkedBox.closest('.dt-cell');
			
			if (!selectedCell) {
				frappe.msgprint(__("Unable to identify the selected row."));
				return false;
			}
		
			// Get the row index from data attribute
			let selectedRowIndex = selectedCell.getAttribute('data-row-index');
		
			if (selectedRowIndex === null) {
				frappe.msgprint(__("Could not fetch the selected row's index."));
				return false;
			}
		
			// Fetch the selected row data using the row index
			let selectedRowData = report.data[selectedRowIndex];P
		
			if (!selectedRowData) {
				frappe.msgprint(__("Unable to fetch data for the selected row."));
				return false;
			}
			
			console.log(selectedRowIndex, selectedRowData);
			
			// Open the Manual Punch form with the selected row's data
			frappe.new_doc("Manual Punch", {
				doctype: "Manual Punch",
				error_date: selectedRowData.attendance_date,
				employee: selectedRowData.employee
			});
		}).addClass("btn-second");
		
		
		
			
	},
	get_datatable_options(options) {
		console.log(options);
		
        return Object.assign(options, {
            checkboxColumn: true
        });
    }
};

File: gke_customization/gke_hrms/report/employee_punch_error/employee_punch_error.py
# Copyright (c) 2024, Frappe Technologies Pvt. Ltd. and contributors
# For license information, please see license.txt

import frappe
from frappe import _

def execute(filters=None):
	columns = get_columns(filters)
	data = get_data(filters)
	return columns, data

def get_data(filters=None):
	conditions = get_conditions(filters)

	if conditions:
		data = frappe.db.sql(f"""SELECT 
				at.employee, at.employee_name, 
				at.company, at.shift,
				at.department,at.attendance_date, 
				# at.name,  tec.name as emp,
				COUNT(tec.name) AS punch_count
			FROM 
				`tabAttendance` AS at
			LEFT JOIN 
				`tabEmployee Checkin` AS tec ON tec.attendance = at.name
			{conditions}
			GROUP BY 
				at.employee, at.attendance_date
			ORDER BY 
				at.attendance_date ASC;
		""", as_dict=1)
	# else:
	# 	data = frappe.db.sql(f"""
	# 		SELECT 
	# 			at.employee, at.employee_name, at.company, at.shift,
	# 			at.department,at.attendance_date, at.name,  tec.name as emp,
	# 			COUNT(tec.name) AS checkin_count
	# 		FROM 
	# 			`tabAttendance` AS at
	# 		LEFT JOIN 
	# 			`tabEmployee Checkin` AS tec ON tec.attendance = at.name
	# 		WHERE 
	# 			(at.in_time IS NOT NULL OR at.out_time IS NOT NULL)
	# 			AND at.out_time IS NULL
	# 			AND at.attendance_date BETWEEN '2025-01-01' AND '2025-01-11'
	# 			AND at.company = 'Gurukrupa Export Private Limited'
	# 			# AND at.department = 'Information Technology - GEPL'
	# 		GROUP BY 
	# 			at.employee, at.attendance_date
	# 		ORDER BY 
	# 			at.attendance_date ASC;
	# 	""", as_dict=1)
	
	# frappe.throw(str(data))
	return data

def get_columns(filters=None):
	columns = [
		{
			"label": _("Company"),
			"fieldname": "company",
			"fieldtype": "Data",
			"width": 250
		},
		{
			"label": _("Employee ID"),
			"fieldname": "employee",
			"fieldtype": "Data",
			"width": 150
		},
		{
			"label": _("Employee Name"),
			"fieldname": "employee_name",
			"fieldtype": "Data",
			"width": 250
		},
		{
			"label": _("Punch Date"),
			"fieldname": "attendance_date",
			"fieldtype": "Date"
		},
		{
			"label": _("Department"),
			"fieldname": "department",
			"fieldtype": "Data",
			"width": 200
		},		
		# {
		# 	"label": _("Punch Times"),
		# 	"fieldname": "time",
		# 	"fieldtype": "Data",
		# 	"hidden":1
		# },
		{
			"label": _("Punch Count"),
			"fieldname": "punch_count",
			"fieldtype": "Data",
			"width": 150
		},
	]
	return columns

def get_conditions(filters):

	filter_list = []

	if filters.get("company"):
		filter_list.append(f'''at.company = "{filters.get("company")}"''')
	
	if filters.get("from_date") and filters.get("to_date"):
		filter_list.append(f'''at.attendance_date Between "{filters.get("from_date")}" and "{filters.get("to_date")}" ''')

	if filters.get("department"):
		filter_list.append(f'''at.department = "{filters.get("department")}"''')

	if filters.get("employee"):
		filter_list.append(f'''at.employee = "{filters.get("employee")}"''')
	

	conditions = "where (at.in_time IS NOT NULL OR at.out_time IS NOT NULL) AND at.out_time IS NULL and " + " and ".join(filter_list)
	if conditions!='where ':
		return conditions